### ISSUE TITLE
Event Handler Middleware

### ISSUE THREAD

**Issue by philbates** [laravel/internals#113]
Imagine the case where you have an optional module that can be enabled or disabled in the application UI. This module registers many event listeners, but you would only want each of these listeners to fire when the module is enabled.

Right now, the cleanest way to do this, as far as I can see, would be something like this:

```
abstract class BlogEventListener
{
    public function handle($event)
    {
        if (!$this->blogEnabled()) {
            return;
        }

        $this->handleBlogEnabled($event);
    }

    abstract function handleBlogEnabled($event);

    protected function blogEnabled()
    {
        return Settings::get('blog_enabled');
    }
}
```

```
class PostBlogWhenNewUserRegisters extends BlogEventListener
{
    public function handleBlogEnabled($newUserRegisteredEvent)
    {
        Blog::create(['title' => "Welcome, {$newUserRegisteredEvent->username}"]);
    }
}
```

Although this would work, IMO it doesn't feel exactly elegant? Now imagine your app has 10-20 different optional modules, each of which has lots of event listeners, that can be enabled or disabled by the UI. That's a lot of events that have to extend abtract handler classes, and just generally a lot of excess noise.

However, then I thought - this is the exact problem that HTTP middleware solves. How about if I could do something like this:

```
class BlogServiceProvider extends ServiceProvider
{
    public function boot()
    {
        // Single middleware...
        Event::listen(NewUserRegistered::class, [
            'middleware' => BlogEnabledEventMiddleware::class,
            'uses' => PostBlogWhenNewUserRegisters::class,
        ]);

        // It would support multiple middlewares
        Event::listen(NewUserRegistered::class, [
            'middleware' => [
                BlogEnabledEventMiddleware::class,
                SomeOtherEventMiddleware::class
            ],
            'uses' => PostBlogWhenNewUserRegisters::class,
        ]);

        // Closure-based middleware
        Event::listen(NewUserRegistered::class, [
            'middleware' => function ($newUserRegisteredEvent) {
                return Settings::get('blog_enabled');
            },
            'uses' => PostBlogWhenNewUserRegisters::class,
        ]);
    }
}
```

```
class BlogEnabledEventMiddleware
{
    public function handle($event, $next)
    {
        if (!$this->blogEnabled()) {
            return;
        }

        return $next($event);
    }

    protected function blogEnabled()
    {
        return Settings::get('blog_enabled');
    }
}
```

Meaning my event listener would become lovely and clean (with bonus type hinting, which wasn't possible in the first examples above), and will only ever be called when the middleware assigned to it decide it should be called.

```
class PostBlogWhenNewUserRegisters
{
    public function handle(NewUserRegistered $newUserRegisteredEvent)
    {
        Blog::create(['title' => "Welcome, {$newUserRegisteredEvent->username}"]);
    }
}
```

Even better, there would be an uncanny symmetry between my route registrations and my event listeners, because I would also put my route registrations behind a HTTP middleware that checks if the blog is enabled:

```
class BlogServiceProvider extends ServiceProvider
{
    public function boot()
    {
        Route::get('blog/{id}', [
            'middleware' => Http\Middleware\BlogEnabled::class,
            'uses' => 'BlogController@index'
        ]);
        Event::listen(NewUserRegistered::class, [
            'middleware' => Events\Middleware\BlogEnabled::class,
            'uses' => PostBlogWhenNewUserRegisters::class,
        ]);
```

The only difference between the HTTP middleware and the event middleware would be that the HTTP middleware would return a 404 response (for example) when the blog is disabled, and the event middleware would just return early and not pass on the event to the next layer.

I took a stab at implementing this myself, and I got so far... but then I realised there's further implications with queued events, and then I started feeling a little lost. Also, I noticed that currently we can't use Illuminate\Pipeline in Illuminate\Events as it isn't included in Illuminate\Events's composer.json, and ideally to implement a middleware system you'd probably want to use the pipeline.

That said, I still think this could be a really useful feature, and if anyone fancies having a go at implementing this, I'd be interested to see what you come up with.

**Comment by franzliedke** [comment]
I'm not 100% sure whether I understand all the intricacies of your situation, but why don't you just set up your event listeners only if the setting is enabled? That way, you'd have that condition in exactly one place, the listeners would not have to deal with it, and you also have a lot less event handlers in the case that the setting is disabled...

**Comment by philbates35** [comment]
Well that was going to be my plan originally, but after a lot of time research the best ways to handle optional modules, there's a few reasons why I've come to the conclustion that always registering the listeners is better:

- My settings (is blog enabled, true or false) are stored in the database (and then cached), and so that would mean having to do a database query and talk to the cache in the `boot` method. This means that the application can't boot without access to a database, and all my tests have to have to use `DatabaseMigrations`. Not the end of the world, but I feel this isn't ideal.

- Imagine at the start of the request the blog module is disabled, then mid-way through the request the setting is enabled by the user, and then before the end of the request an event is fired that the blog module listens to, I would expect the blog event handler to be triggered. However, if I only registered the event listener at boot time if the setting was enabled, then this wouldn't work because the listener would only be registered from the next request onwards, whereas with the middleware solution, the check to see if the module is enabled or not or deferred until the event is fired, as now the blog is enabled the handler would be triggered. Granted, this is an edge case and almost certainly would never happen - but theoretically it could happen, and when it did, I think this behavior would not be intuative.

- The more I think about it, the answer to your question "why not just conditionally register the listeners?" is the same as the answer to "why don't we conditionally register route?". For example, you wouldn't do:
    ```
    public function boot()
    {
        if (Auth::user()->superAdmin()) {
            Route::get('admin', 'AdminController@index');
        }
    }
    ```
    Of course, instead you'd do:
    ```
    public function boot()
    {
        Route::get('admin', ['middleware' => 'superadmin', 'uses' => 'AdminController@index']);
    }
    ```

The crux of the issue to me is that it feels inconsistent to me that you would always register routes related to the optional module and add a middleware there, but conditionally register event listeners. Either both routes and events should be conditionally registered, or always registered with a "middleware" check just before it is triggered. The lack of "event middleware" at this point makes this impossible.

**Comment by philbates35** [comment]
That said... I am realising that although this "event middleware" solution would work for events, and I still believe this is a feature worth implementing, I would still have the same problem with view composers, macros, and any extensions to my application services (basically anything that would go it `boot()`) that are provided by an optional module service provider :/

I still don't like the idea of querying the database in the `boot()` method though. Another reason I'm against this is that I'd like to do the following in my tests:

```
public function test_a_list_a_all_blogs_can_be_viewed_when_blog_module_is_enabled()
{
    setting(['blog_enabled' => true]);

    $this->call('GET', 'blogs');

    $this->assertResponseStatus(200);
}

public function test_a_404_is_returned_when_blog_module_is_disabled()
{
    setting(['blog_enabled' => false]);

    $this->call('GET', 'blogs');

    $this->assertResponseStatus(404);
}
```
If I conditionally register stuff in boot based on the settings in the database, then by the time my test starts (after the app has booted) my call to set the relevant setting using `setting()` will have no affect because the app has already booted and one of those two tests will fail.

**Comment by franzliedke** [comment]
I don't think the comparison with routes is valid:

When you register a route, a callback is stored, but not executed. Only one route is ever executed per request.

When you register an event handler, a callback is stored. Several event handlers can be executed per request. If you have a condition inside them, they will still be executed up to that condition.

**Comment by franzliedke** [comment]
As for querying the database, you could, of course, create some kind of composite event listener that checks the settings and then, depending on the result, immediately subscribes and executes the other listeners.

Advantages:
- you only execute the query when you need it
- you get rid of the duplication in the various listener classes
- concerns are nicely separated
- it works with the current system

**Comment by philbates35** [comment]
So after a couple of days experimenting, I came up with the following that sounds similar to what you are suggesting. Basically, at the time I register the listener, I at the same time pass a closure that acts as a filter. I wrap the provided handler in a closure that will act as the real handler. The real handler, calls the filter, and only if the filter doesn't return false does it then call the handler that was originally registered. Note that if the original event handler was a class-based handler, that handler class will not be resolved out of the container until the filter passes.

That last part is important - The handler class (e.g. `BlogEventListener`) will probably have other dependencies from my optional module, e.g. `BlogService`. If I inject `BlogService` into the handler and the handler is called right away as is the behaviour out of the box, the `BlogService` and all of its other blog-related dependencies (`AnotherBlogService` and `BlogRepository`) will be all be resolved out of the container. This is okay if the blog module is enabled, but if the blog module isn't enabled, then it's not right that all those blog-services are resolved out of the container when the event is fired.

With that in mind, here's my solution so far. I'd be interested to hear any feedback or improvements you think i could make.

```php
trait ListensConditionally
{
    /**
     * Listen to an event, and only call the listener when the result of the
     * filter callback that is called first doesn't return false. If it
     * returns anything other than false, the registered listener will
     * be called.
     *
     * @param  string|array  $events
     * @param  mixed  $listener
     * @param  \Closure $filter
     * @param  int  $priority
     * @return void
     */
    public function listenConditionally($events, $listener, Closure $filter, $priority = 0)
    {
        $dispatcher = $this->app['events'];

        $conditionalListener = function () use ($dispatcher, $listener, $filter) {
            $payload = func_get_args();

            if (call_user_func_array($filter, $payload) === false) {
                return;
            }

            return call_user_func_array($dispatcher->makeListener($listener), $payload);
        };

        $dispatcher->listen($events, $conditionalListener, $priority);
    }
}
```
```php
use Closure;
use Illuminate\Contracts\Events\Dispatcher as DispatcherContract;
use Illuminate\Support\ServiceProvider;

abstract class ConditionalEventServiceProvider extends ServiceProvider
{
    use ListensConditionally;

    /**
     * The event handler mappings for the application.
     *
     * @var array
     */
    protected $listen = [];

    /**
     * Register the application's event listeners.
     *
     * @param  \Illuminate\Contracts\Events\Dispatcher  $events
     * @return void
     */
    public function boot(DispatcherContract $events)
    {
        foreach ($this->listens() as $event => $listeners) {
            foreach ($listeners as $listener) {
                $this->listenConditionally($event, $listener, $this->filterClosure());
            }
        }
    }

    /**
     * Register the service provider.
     *
     * @return void
     */
    public function register()
    {
        //
    }

    /**
     * Get the events and handlers.
     *
     * @return array
     */
    public function listens()
    {
        return $this->listen;
    }

    /**
     * Return the closure that will be used as a filter before
     * firing the event listeners.
     *
     * @return \Closure
     */
    protected function filterClosure()
    {
        return function () {
            return true;
        };
    }
}
```
This means that my blog event service provider looks like nice and clean:
```php
<?php

use Illuminate\Contracts\Events\Dispatcher as DispatcherContract;

class EventServiceProvider extends ConditionalEventServiceProvider
{
    protected $listen = [
        \Some\Event::class => [
            \Handler\Instantiated\Only\When\Filter\Returns\True::class,
        ],
    ];

    public function boot(DispatcherContract $events)
    {
        parent::boot($events);

        $this->listenConditionally(
            \Some\Event\That\Needs\Priority::Class,
            \App\Listeners\FooListener::class,
            $this->filterClosure(),
            100
        );
    }

    protected function filterClosure()
    {
        return function () {
            return setting('biometrics_enabled');
        };
    }
}
```
To summarise, the benefits of this are:
- No duplication of `if (setting('blog_enabled'))` at the start of every `handle` method
- The handler class, and therefore its dependencies are only resolved out of the container only once it is certain that the blog module is enabled. When it isn't enabled and the event is fired, no blog-related services or classes are resolved at all, or the handler itself, are resolved.

**Comment by franzliedke** [comment]
Sounds like a very good solution. What keeps it from being perfect? 

I am not so sure we need this in core. Let's hear what others (and Taylor) say.

**Comment by philbates35** [comment]
Yeah to be honest the more I think about it the more you're probably right that this isn't needed in the core. I guess most web apps won't need this level of modular configurablity and everything that comes with it, such as event listeners that should sometime trigger and sometimes shouldn't.

Glad to hear you think it's a good solution. :)
